src/modules/julia.rs

```diff
@@ -0,0 +1,558 @@
+        // wait user input
+        csleep(rng.gen_range(500..2500)).await;
+
+        // enter pkg mode
+        print_pkg_prompt(project, true).await;
+
+        // wait user input
+        csleep(rng.gen_range(500..1500)).await;
+
+        // type "update" and press enter
+        dprint("up", rng.gen_range(100..500)).await;
+        dprint("date", rng.gen_range(100..500)).await;
+        csleep(rng.gen_range(500..1500)).await;
+        newline().await;
+
+        // wait Pkg.update() startup time
+        csleep(rng.gen_range(200..1000)).await;
+
+        log_action("Updating", "registry at `~/.julia/registries/General.toml`").await;
+        csleep(rng.gen_range(1500..5000)).await;
+
+        log_action("Resolving", "package versions...").await;
+        csleep(rng.gen_range(500..2500)).await;
+
+        install_packages(packages).await;
+
+        csleep(rng.gen_range(250..1000)).await;
+
+        download_artifacts(artifacts).await;
+
+        csleep(rng.gen_range(250..1000)).await;
+
+        update_project_and_manifest(project).await;
+
+        csleep(rng.gen_range(10..100)).await;
+
+        report_packages(packages).await;
+
+        csleep(rng.gen_range(150..500)).await;
+
+        build_artifacts(artifacts).await;
+
+        csleep(rng.gen_range(150..500)).await;
+
+        precompile(packages).await;
+
+        if rng.gen::<f32>() < 0.25f32 {
+            gc().await;
+        }
+
+        csleep(rng.gen_range(50..250)).await;
+
+        print_pkg_prompt(project, false).await;
+
+        // wait user input
+        csleep(rng.gen_range(500..5000)).await;
+
+        // exit pkg mode
+        print_julia_prompt(true).await;
+
+        // wait user input
+        csleep(rng.gen_range(1000..7000)).await;
+
+        // quit julia
+        newline().await;
+
+        if appconfig.should_exit() {
+            return;
+        }
+    }
+}
+
+async fn print_banner() {
+    print(format!(
+        r#"               {gu}
+   {bu}       _ {ru}{g}{mu}     |  Documentation: https://docs.julialang.org
+  {b}     | {r} {m}    |
+   _ _   _| |_  __ _   |  Type "?" for help, "]?" for Pkg help.
+  | | | | | | |/ _` |  |
+  | | |_| | | | (_| |  |  Version 1.7.3 (2022-05-06)
+ _/ |\__'_|_|_|\__'_|  |  Fedora 35 build
+|__/                   |
+"#,
+        r = Paint::red("(_)").bold(),
+        ru = Paint::red("_").bold(),
+        g = Paint::green("(_)").bold(),
+        gu = Paint::green("_").bold(),
+        b = Paint::blue("(_)").bold(),
+        bu = Paint::blue("_").bold(),
+        m = Paint::magenta("(_)").bold(),
+        mu = Paint::magenta("_").bold(),
+    ))
+    .await;
+}
+
+async fn print_julia_prompt(overwrite: bool) {
+    if overwrite {
+        erase_line().await;
+    } else {
+        newline().await;
+    }
+    print(format!("{} ", Paint::green("julia>").bold())).await;
+}
+
+async fn print_pkg_prompt(project: &str, overwrite: bool) {
+    if overwrite {
+        erase_line().await;
+    } else {
+        newline().await;
+    }
```

Comment:
Instead of taking the `overwrite` parameter here and everywhere else, wouldn't it perhaps be nicer to just either `erase_line()` or `newline()` just before you call this function (and others that take `overwrite`)? I think it would be more concise and also more obvious.

Reply:
I believe this reflects more closely the actual behavior of the Julia prompt.

- Let's say that the session is in this state (the `│` denotes the cursor position)
  ```
  julia> 5│
  ```
  and you press enter. Then Julia prints the answer followed by a newline `5\n`
  ```
  julia> 5
  5
  │
  ```
  and then prints the new prompt `\njulia> ` (which has an additional newline at the start)
  ```
  julia> 5
  5
  
  julia> │
  ```

- You can convince yourself that the prompt has a newline at the start by looking at this example. Let's say that the session is in this state
  ```
  julia> print(5)│
  ```
  and you press enter. Then Julia prints just `5`, without a newline
  ```
  julia> 5
  5│
  ```
  and then prints the new prompt `\njulia> ` (which has an additional newline at the start), resulting in
  ```
  julia> 5
  5
  julia> │
  ```

- If instead the session is in this state
  ```
  julia> │
  ```
  and you type `]`, then Julia replaces the prompt with the new one (by first erasing the line)
  ```
  (@v1.7) pkg> │
  ```

For this reason, I decided to implement `print_*_prompt(..., overwrite=false)` for typing a new prompt and `print_*_prompt(..., overwrite=true)` for replacing the current prompt.

Reply:
If the `overwrite: bool` argument can be confusing, I can replace it with something like

```rust
enum PromptPrinting {
    New,
    Replace
}

use PromptPrinting::*;

async fn print_julia_prompt(printing: PromptPrinting) {
    match printing {
        New => newline().await;
    } else {
        Replace => erase_line().await;
    }
    print(format!("{} ", Paint::green("julia>").bold())).await;
}
```

which can then be used as

```rust
        // julia startup
        print_banner().await;
        csleep(rng.gen_range(50..150)).await;
        print_julia_prompt(New).await;             // <-- change

        // wait user input
        csleep(rng.gen_range(500..2500)).await;

        // enter pkg mode
        print_pkg_prompt(project, Replace).await;  // <-- change
```

Reply:
I think we might be miscommunicating. I believe the current output is sound but that the code condition is not something we actually need to get there. Currently you do this:
```rust
        // enter pkg mode
        print_pkg_prompt(project, true).await;
        ...
        print_pkg_prompt(project, false).await;
```
How about instead we do:
```rust
        erase_line().await;
        print_pkg_prompt(project).await;
        ...
        newline().await;
        print_pkg_prompt(project).await;
```
However, I think you are arguing that you want to model the functions the same way the `julia` output works (as opposed to my idea of having a simpler but less real-world modelling approach). I'm actually fine with either but I usually prefer simplicity over modelling.

Reply:
> However, I think you are arguing that you want to model the functions the same way the `julia` output works (as opposed to my idea of having a simpler but less real-world modelling approach). I'm actually fine with either but I usually prefer simplicity over modelling.

Yes, I was trying to do that, but I do agree that it's overkill for this simple situation. I'll rewrite it the way you prefer, since it's your project after all and I'm only a guest 🙂 — *When in Rome, do as the Romans do*.